mlir/lib/Transforms/Utils/DialectConversion.cpp

```diff
@@ -63,80 +113,140 @@ static OpBuilder::InsertPoint computeInsertPoint(Value value) {
   return OpBuilder::InsertPoint(insertBlock, insertPt);
 }
 
+/// Helper function that computes an insertion point where the given values are
+/// defined and can be used without a dominance violation.
+static OpBuilder::InsertPoint computeInsertPoint(ArrayRef<Value> vals) {
+  assert(!vals.empty() && "expected at least one value");
+  OpBuilder::InsertPoint pt = computeInsertPoint(vals.front());
+  for (Value v : vals.drop_front())
+    pt = chooseLaterInsertPoint(pt, computeInsertPoint(v));
+  return pt;
+}
+
 //===----------------------------------------------------------------------===//
 // ConversionValueMapping
 //===----------------------------------------------------------------------===//
 
+/// A vector of SSA values, optimized for the most common case of a single
+/// value.
+using ValueVector = SmallVector<Value, 1>;
+
 namespace {
+
+/// Helper class to make it possible to use `ValueVector` as a key in DenseMap.
+struct ValueVectorMapInfo {
+  static ValueVector getEmptyKey() { return ValueVector{}; }
+  static ValueVector getTombstoneKey() { return ValueVector{}; }
```

Comment:
I think these values are problematic, could we use `ValueVector{DenseMapInfo<Value>::getEmptyKey()}` instead?
Otherwise 1:0 type conversions would not work I believe. Is it easy to add a test for this?

Reply:
Side note: 1:0 conversions don't go through the `ConversionValueMapping`. We immediately add a source materialization in that case (look for `// Materialize a replacement value "out of thin air".` in the code base). So we actually store a 1:1 mapping in that case. (We have test cases for that.)